Sources/StreamChat/Workers/ChannelListUpdater.swift

```diff
@@ -51,23 +52,34 @@ class ChannelListUpdater: Worker {
                         
                         let localQueryCIDs = Set(queryDTO.channels.compactMap { try? ChannelId(cid: $0.cid) })
                         let remoteQueryCIDs = Set(channelListPayload.channels.map(\.channel.cid))
-                        let queryAlreadySynched = remoteQueryCIDs.intersection(synchedChannelIds)
-                        
-                        // We are going to unlink & clear those channels that are not present in the remote query,
-                        // and that have not been synched nor watched. Those are outdated.
-                        let cidsToRemove = localQueryCIDs
-                            .subtracting(queryAlreadySynched)
-                            .subtracting(watchedChannelIds)
-                        
-                        for cid in cidsToRemove {
+
+                        let updatedChannels = synchedChannelIds.union(watchedChannelIds)
+                        let localNotInRemote = localQueryCIDs.subtracting(remoteQueryCIDs)
+                        let localInRemote = localQueryCIDs.intersection(remoteQueryCIDs)
+
+                        // We unlink those local channels that are no longer in remote
+                        for cid in localNotInRemote {
                             guard let channelDTO = session.channel(cid: cid) else { continue }
-                            
-                            channelDTO.resetEphemeralValues()
-                            channelDTO.messages.removeAll()
                             queryDTO.channels.remove(channelDTO)
                         }
+
+                        // We are going to clean those channels that are present in the both the local and remote query,
+                        // and that have not been synched nor watched. Those are outdated, can contain gaps.
+                        let cidsToClean = localInRemote.subtracting(updatedChannels)
+                        session.cleanChannels(cids: cidsToClean)
```

Comment:
I would only make `RefetchChannelListQueryOperation` operation responsible for unlinking channels from a query and move channels clean-up fully to `CleanUnwantedChannelsOperation` as it makes easier to understand what's happening.

If there's a channel `C1` locally and it is linked to `Q1` and `Q2` local queries, then when the 1st page for `Q1` is fetched `C1` can be outside this page but can still be part of `Q2's` first page. 

I see that we're subtracting `updatedChannels` which is a union of synced & watched but it's easier to make a decision which channels are/can get outdated when all queries are refetched.

Reply:
What if we just clean all channels that are not linked to any query after connection recovery subtracting watched channels? 🤔 because the math is too complex, I'm already lost with it 😄

Reply:
There are 2 things here:

> I would only make `RefetchChannelListQueryOperation` operation responsible for unlinking channels
It is something needed to leave that particular channel list ready for it to be used. I don't think the responsibility should be offloaded to the operation. Actually, the operation should only know about the result of it, IMO.

> What if we just clean all channels that are not linked to any query after connection recovery subtracting watched channels? 🤔 because the math is too complex, I'm already lost with it 😄
Happy to receive suggestions here. What I can tell you now is that I tried many things, and this was the least complex while still readable that I managed to do 😂

Reply:
> It is something needed to leave that particular channel list ready for it to be used

Yes, but we can't make a decision to clean up a local channel matching a query when it's outside the 1st page from remote as it can be on the 1st page of another query and would get synced in the end if we didn't clean it up.

Reply:
> Happy to receive suggestions here. What I can tell you now is that I tried many things, and this was the least complex while still readable that I managed to do 😂

If we make only `CleanUnwantedChannelsOperation` responsible for cleaning channels, the operation key par could look like that:
```swift
let channelsToClean = context
  .localChannelIds
  .compactMap { session.channel(cid: $0) }
   // clean a channel that is not linked to any query excluding watched channels
  .filter { $0.queries.isEmpty && !context.watchedChannelIds.contains($0.cid) }
```swift

The code above could be done with 1 fetch request with a compound predicate to avoid multiple fetches to the persistent store for each single channel

Reply:
> Yes, but we can't make a decision to clean up a local channel matching a query when it's outside the 1st page from remote as it might be on the 1st page of another query and might get synced in the end if we didn't clean it up. CleanUnwantedChannelsOperation seems to be designed for that very purpose

We actually can because at that step we are only cleaning those which are present in the remote query, which means they will get populated with new data.

> If we make only CleanUnwantedChannelsOperation responsible for cleaning channels, the operation key part could look like that:

IMO it is better to have the current logic than involve fetching all the channels as you suggested above

Reply:
Discussed on the call ✅ Pol was saying about local channels that are on remote but not synced - we can clean them as part of query re-fetch operation. I was saying about channels that are outside the 1st page that will go out of sync if not watched. Closing this as we're on the same page now 🙂